File: app/types/safety-settings.ts
import type { Environment } from '@/types'

/**
 * 確認ダイアログを表示する危険度レベルの閾値
 */
export type ConfirmationThreshold = 'warning' | 'danger'

/**
 * 環境別の安全設定
 */
export interface EnvironmentSafetyConfig {
  /** 確認ダイアログを有効にするか */
  confirmationEnabled: boolean

  /** 確認ダイアログを表示する危険度の閾値 */
  confirmationThreshold: ConfirmationThreshold

  /** カウントダウン秒数（0〜10） */
  countdownSeconds: number

  /** DROPクエリを禁止するか */
  disableDrop: boolean

  /** TRUNCATEクエリを禁止するか */
  disableTruncate: boolean
}

/**
 * 全環境の安全設定
 */
export interface SafetySettings {
  version: number
  environments: Record<Environment, EnvironmentSafetyConfig>
}

/**
 * デフォルト設定
 */
export const DEFAULT_SAFETY_SETTINGS: SafetySettings = {
  version: 1,
  environments: {
    development: {
      confirmationEnabled: true,
      confirmationThreshold: 'danger',
      countdownSeconds: 0,
      disableDrop: false,
      disableTruncate: false,
    },
    test: {
      confirmationEnabled: true,
      confirmationThreshold: 'danger',
      countdownSeconds: 0,
      disableDrop: false,
      disableTruncate: false,
    },
    staging: {
      confirmationEnabled: true,
      confirmationThreshold: 'warning',
      countdownSeconds: 3,
      disableDrop: false,
      disableTruncate: false,
    },
    production: {
      confirmationEnabled: true,
      confirmationThreshold: 'warning',
      countdownSeconds: 5,
      disableDrop: true,
      disableTruncate: true,
    },
  },
}

File: src-tauri/src/models/safety_settings.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum ConfirmationThreshold {
    Warning,
    Danger,
}

impl Default for ConfirmationThreshold {
    fn default() -> Self {
        Self::Warning
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EnvironmentSafetyConfig {
    pub confirmation_enabled: bool,
    pub confirmation_threshold: ConfirmationThreshold,
    pub countdown_seconds: u8,
    pub disable_drop: bool,
    pub disable_truncate: bool,
}

impl Default for EnvironmentSafetyConfig {
    fn default() -> Self {
        Self {
            confirmation_enabled: true,
            confirmation_threshold: ConfirmationThreshold::Warning,
            countdown_seconds: 3,
            disable_drop: false,
            disable_truncate: false,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SafetySettings {
    pub version: u32,
    pub environments: HashMap<String, EnvironmentSafetyConfig>,
}

impl Default for SafetySettings {
    fn default() -> Self {
        let mut environments = HashMap::new();

        environments.insert(
            "development".to_string(),
            EnvironmentSafetyConfig {
                confirmation_enabled: true,
                confirmation_threshold: ConfirmationThreshold::Danger,
                countdown_seconds: 0,
                disable_drop: false,
                disable_truncate: false,
            },
        );

        environments.insert(
            "test".to_string(),
            EnvironmentSafetyConfig {
                confirmation_enabled: true,
                confirmation_threshold: ConfirmationThreshold::Danger,
                countdown_seconds: 0,
                disable_drop: false,
                disable_truncate: false,
            },
        );

        environments.insert(
            "staging".to_string(),
            EnvironmentSafetyConfig {
                confirmation_enabled: true,
                confirmation_threshold: ConfirmationThreshold::Warning,
                countdown_seconds: 3,
                disable_drop: false,
                disable_truncate: false,
            },
        );

        environments.insert(
            "production".to_string(),
            EnvironmentSafetyConfig {
                confirmation_enabled: true,
                confirmation_threshold: ConfirmationThreshold::Warning,
                countdown_seconds: 5,
                disable_drop: true,
                disable_truncate: true,
            },
        );

        Self {
            version: 1,
            environments,
        }
    }
}
